kentik_api_library/kentik_api/synthetics/synth_tests/base.py

```diff
@@ -41,55 +43,119 @@ class Defaults:
+            if isinstance(src_value, list):
+                if not src_value:
+                    return []
+                list_item_type = type(src_value[0])  # assuming either primitive type or type with "to_pb()" method
+                return [get_value(list_item_type, e) for e in src_value]
+            if isinstance(src_value, dict):
+                if not src_value:
+                    return {}
+                dict_value_type = type(list(src_value.values())[0])  # assuming primitive type or type with "to_pb()"
+                return {k: get_value(dict_value_type, v) for k, v in src_value.items()}
+            try:
+                return dst_type(src_value)
+            except TypeError as error:
+                raise RuntimeError(f"Don't know how to serialize to '{dst_type}' (value: '{src_value}')") from error
+
+        dummy_pb_instance = self.PB_TYPE()  # used only for examining the destination protobuf object field types
+        args = {}
+        for f in fields(self):
+            dst_field_name = f.name[1:] if (f.name[0] == "_") else f.name  # handle private fields that start with "_"
+            dst_field_type = type(getattr(dummy_pb_instance, dst_field_name))
+            src_field_value = getattr(self, f.name)
+            args[dst_field_name] = get_value(dst_field_type, src_field_value)
+        return self.PB_TYPE(**args)
 
+    @classmethod
+    def from_pb(cls: Type[_ConfigElementT], obj: Any) -> _ConfigElementT:
+        def get_value(dst_type: Type[Any], src_value: Any) -> Any:
+            if hasattr(dst_type, "from_pb"):
+                return dst_type.from_pb(src_value)
             try:
-                return f(v)
+                return dst_type(src_value)
             except TypeError as error:
-                if f._name == "List":
-                    return [get_value(type(i), i) for i in v]
-                if f._name == "Dict":
-                    return {_k: get_value(type(_v), _v) for _k, _v in v.items()}
-                raise RuntimeError(f"Don't know how to instantiate '{f}' (value: '{v}')") from error
-
-        _init_fields = {f.name: f for f in fields(cls) if f.init}
-        args = {k: get_value(_init_fields[k].type, v) for k, v in d.items() if k in _init_fields.keys()}
-        # noinspection PyArgumentList
-        inst = cls(**args)  # type: ignore
-        _other_fields = {f.name: f for f in fields(cls) if not f.init}
-        for n, f in _other_fields.items():
-            if n[0] == "_":
-                k = n.split("_")[1]
-            else:
-                k = n
-            if k in d:
-                setattr(inst, n, get_value(f.type, d[k]))
-        return inst
+                if dst_type._name == "List":
+                    list_item_type = get_args(dst_type)[0]
+                    return [get_value(list_item_type, i) for i in src_value]
+                if dst_type._name == "Dict":
+                    dict_value_type = get_args(dst_type)[1]
```

Comment:
This almost certainly won't work `Union`  and `Any` type values (will fail with the `RuntimeError` bellow), but that is probably not a problem, assuming that the code works :). Do we actually have instances of dictionaries whose value types differ in the protobuf compared to the target object? If not, we may as well use `type(_v)` bellow.

Reply:
The error with Union and Any doesn't occur as we don't use these types in dataclasses.
type(_v) won't work for TraceResponse where we have Dict[str, NetNode].